test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportGenerator.cs

```diff
@@ -66,89 +76,134 @@ public async Task<ITestResultReport> CreateReportAsync()
             ulong mismatchSuccess = 0;
             ulong mismatchFailure = 0;
 
-            bool hasExpectedResult = await this.ExpectedTestResults.MoveNextAsync();
-            bool hasActualResult = await this.ActualTestResults.MoveNextAsync();
+            bool hasSenderResult = await this.SenderTestResults.MoveNextAsync();
+            bool hasReceiverResult = await this.ReceiverTestResults.Match(async x => await x.MoveNextAsync(), () => Task.FromResult(false));
 
-            while (hasExpectedResult && hasActualResult)
+            while (hasSenderResult)
             {
-                this.ValidateDataSource(this.ExpectedTestResults.Current, this.ExpectedSource);
-                this.ValidateDataSource(this.ActualTestResults.Current, this.ActualSource);
-
+                this.ValidateDataSource(this.SenderTestResults.Current, this.SenderSource);
                 (NetworkControllerStatus networkControllerStatus, bool isWithinTolerancePeriod) =
-                        this.NetworkStatusTimeline.GetNetworkControllerStatusAndWithinToleranceAt(this.ExpectedTestResults.Current.CreatedAt);
-                if (!NetworkControllerStatus.Enabled.Equals(networkControllerStatus) &&
-                    !NetworkControllerStatus.Disabled.Equals(networkControllerStatus))
+                    this.NetworkStatusTimeline.GetNetworkControllerStatusAndWithinToleranceAt(this.SenderTestResults.Current.CreatedAt);
+                this.ValidateNetworkControllerStatus(networkControllerStatus);
+                DirectMethodTestResult dmSenderTestResult = JsonConvert.DeserializeObject<DirectMethodTestResult>(this.SenderTestResults.Current.Result);
+                if (hasReceiverResult)
                 {
-                    throw new InvalidOperationException($"Unexpected Result. NetworkControllerStatus was {networkControllerStatus}");
+                    string receiverSource = this.ReceiverSource.Expect<ArgumentException>(
+                        () => throw new ArgumentException("There must be a receiver source if there are receiver test results"));
+                    ITestResultCollection<TestOperationResult> receiverTestResults = this.ReceiverTestResults.Expect<ArgumentException>(
+                        () => throw new ArgumentException("Previous check for receiver source passed"));
+                    this.ValidateDataSource(receiverTestResults.Current, receiverSource);
+                    if (!this.TestResultComparer.Matches(receiverTestResults.Current, this.SenderTestResults.Current))
```

Comment:
you need to think about it may be better to deserialize directly here instead of use comparer.

Reply:
Let's do pairing together tomorrow.

Reply:
Serializing directly. Got rid of comparer